src/NewKeyManager.sol

```diff
@@ -0,0 +1,186 @@
+
+abstract contract NewKeyManagerBase {
+    // This base class provides the functionality of a singleton
+    // Private Key holder. It allows many applications to share the
+    // same bootstrapped instance.
+
+    // Anyone can see the master public key
+    address public xPub;
+
+    // placeholder for hardened and non_hardened child indexes
+    uint32 public constant HARDENED_START_INDEX = 2147483648;
+
+    // Attestation is now possible using the caller as domain
+    // separator
+    function attest(bytes32 appData) public returns (bytes memory) {
+        bytes32 digest = keccak256(abi.encodePacked(msg.sender, appData));
+        
+        return Secp256k1.sign(uint256(xPriv()), digest);
+    }
+
+    function verify(address caller, bytes32 appData, bytes memory sig) public view returns (bool) {
+        bytes32 digest = keccak256(abi.encodePacked(caller, appData));
+        return Secp256k1.verify(xPub, digest, sig);
+    }
+
+    //////////////////////////////////
+    // To be implemented by subclasses
+    //////////////////////////////////
+
+    // Only the contract in confidential mode can access the
+    // master private key
+    function seed() internal virtual returns (bytes32);
+
+    function xPriv() internal virtual returns (bytes32);
+
+    function _derivedPriv(address a, uint32 index) internal virtual returns (bytes32);
+
+    // Key derivation for encryption
+
+    // Because we are using hardened derivation, for each
+    // contract we will need someone to sign it off chain
+
+    mapping(bytes32 => bytes) internal _derivedPub;
+
+    function onchain_DeriveKey(
+        address a,
+        bytes memory dPub,
+        // Signature from a valid kettle
+        bytes memory sig,
+        uint32 index
+    ) public {
+        bytes32 digest = keccak256(abi.encodePacked(a, index, dPub));
+        require(Secp256k1.verify(xPub, digest, sig));
+        bytes32 key = keccak256(abi.encodePacked(a, index));
+        _derivedPub[key] = dPub;
+    }
+
+    function offchain_DeriveKey(address a, uint32 index) public returns (bytes memory dPub, bytes memory sig) {
+        bytes32 dPriv = _derivedPriv(a, index);
+        dPub = PKE.derivePubKey(dPriv);
+        bytes32 digest = keccak256(abi.encodePacked(a, index, dPub));
+        sig = Secp256k1.sign(uint256(xPriv()), digest);
+        require(Secp256k1.verify(xPub, digest, sig));
+    }
+}
+
+contract NewKeyManager_v0 is NewKeyManagerBase {
+    IAndromeda public Suave;
+    BIP32 public bip32;
+
+
+    constructor(address _Suave, BIP32 _bip32) {
+        Suave = IAndromeda(_Suave);
+        bip32 = _bip32;
+    }
+
+    function derivedPub(address a, uint32 index) public view returns (bytes memory) {
+        bytes32 key = keccak256(abi.encodePacked(a, index));
+        return _derivedPub[key];
+    }
+
+    function derivedPub(address a) public view returns (bytes memory) {
+        bytes32 key = keccak256(abi.encodePacked(a, uint32(0)));
```

Comment:
This is just a wrapper function for API backward compatibility.
If you use this function, it will return the public key for that address (first layer of the key derivation tree)
See explanation above